File: Demos/demo_mega_scggm.m
addpath('../Mega-sCGGM');

% Simple example
n = 100; p = 200; q = 150;
LambdaSimple = full(spdiags([0.3*ones(q,1) ones(q,1) 0.3*ones(q,1)],[-1 0 1], q, q));
p_influence = floor(p * 0.5);
ThetaSimple = [full(sprand(p_influence, q, 0.01)); zeros(p-p_influence, q)];
X = randn(n, p); % (10 samples x 15 features)
meanY = -X*ThetaSimple*inv(LambdaSimple); 
noiseY = ((chol(LambdaSimple,'lower')')\randn(q, n))';% (10 samples x 12 features)
Y = meanY + noiseY;

lambdaLambda = 0.5;
lambdaTheta = 1;
[estLambdaSimple, estThetaSimple, statsSimple] = mega_scggm(...
    Y, X, lambdaLambda, lambdaTheta);
figure('name', 'Simple demo');
subplot(2,2,1); spy(LambdaSimple); title('true Lambda');
subplot(2,2,2); spy(ThetaSimple); title('true Theta');
subplot(2,2,3); spy(estLambdaSimple); title('est Lambda');
subplot(2,2,4); spy(estThetaSimple); title('est Theta');

% Warm-starting example
figure('name', 'Warm-starting demo');
options.max_outer_iters = 20;
numSuggestions = 5;
[lambdaLambdas, lambdaThetas] = regularization_suggestions(Y, X, ...
    'numSuggestions', numSuggestions);
[estLambda, estTheta, estStats] = mega_scggm(Y, X, lambdaLambdas(1), lambdaThetas(1), options);
subplot(numSuggestions, 2, 1); spy(estLambda);
subplot(numSuggestions, 2, 2); spy(estTheta);

for reg_ix=2:numSuggestions
    lambdaLambda = lambdaLambdas(reg_ix);
    lambdaTheta = lambdaThetas(reg_ix);
    options.Lambda0 = estLambda;
    options.Theta0 = estTheta;
    [estLambda, estTheta] = mega_scggm(Y, X, lambdaLambda, lambdaTheta, options);
    subplot(numSuggestions, 2, (reg_ix-1)*2+1); spy(estLambda);
    subplot(numSuggestions, 2, (reg_ix-1)*2+2); spy(estTheta);
end

File: Demos/demo_mega_scggm.py
#!/bin/python

import math
import matplotlib.pyplot as plt
import numpy as np
import scipy.sparse as ssp
import scipy.sparse.linalg as ssl
import sys

sys.path.append("../Mega-sCGGM/")
from mega_scggm import mega_scggm

n = 100
p = 200
q = 150
Lambda = ssp.eye(q, format="coo")
Lambda.setdiag(0.3, 1)
Lambda.setdiag(0.3, -1)
p_influence = int(math.floor(p * 0.5))
Theta_top = ssp.random(p_influence, q)
Theta_bottom = ssp.coo_matrix((p-p_influence,q))
Theta = ssp.vstack([Theta_top, Theta_bottom])
X = np.random.randn(n, p)
Sigma = ssl.inv(Lambda)
meanY = -1 * X @ Theta @ Sigma
try:
    import sksparse.cholmod as skc
    Lambda_factor = skc.cholesky(Lambda)
    noiseY = (Lambda_factor.solve_Lt(np.random.randn(q,n))).transpose()
except:
    noiseY = np.random.multivariate_normal(np.zeros(q), Sigma.todense(), size=n)
Y = meanY + noiseY

lambdaLambda = 0.5
lambdaTheta = 1.0
(estLambda, estTheta, estStats) = mega_scggm(Y, X, lambdaLambda, lambdaTheta)

plt.figure()
plt.spy(Theta)
plt.title("Theta")

plt.figure()
plt.spy(Lambda)
plt.title("Lambda")

plt.figure()
plt.spy(estLambda)
plt.title("estLambda")

plt.figure()
plt.spy(estTheta)
plt.title("estTheta")


plt.show()
